### DISCUSSION TITLE
Teste unitário e suas boas práticas

### DISCUSSION THREAD

**Discussion by Cleysonlb** [frontendbr/forum#420] [Discussão]
Fala galera, blz?!
Fui designado para começar a escrever os testes unitários aqui na empresa, hoje nosso sistema cobre 0% de teste, o sistema é em ruby on rails e sinceramente, estou me sentindo meio perdido, quais são as boas praticas e casos comuns de teste?!

**Comment by joao-alberto** [comment]
A boa prática é escrever testes. uheuhe 😆  Brinks

Antes de tudo recomendo a leitura desses dois posts, para ficar menos perdido:

https://willianjusten.com.br/entendendo-testes-de-software/
https://nandovieira.com.br/comecando-com-testes-automatizados-no-ruby

Depois disso, fica mais fácil seguir uma trilha de aprendizado. :v:

**Comment by laurindo** [comment]
Sugiro estudar sobre TDD.
Acredito que qualquer prática de teste é um plus gigante. Amigo meu ontem
fez uma entrevista pra trabalhar na Irlanda. A primeira pergunta do CEO foi
saber se ele usava teste unitários em seus projetos.

**Comment by ghost** [comment]
Se quiser um bom curso pra iniciar pense no da [Udacity](https://br.udacity.com/course/javascript-testing--ud549/). É um ótimo curso pois tem ótima didática. Em geral dentro do processo de desenvolvimento de software o teste unitário é importantíssimo visto com bons olhos no mercado. Agora antes de sair testando tudo pende em pesquisar sobre TDD,BDD e etc todos eles são metodologias de teste, cada uma tem características que a tornam diferentes, mas a base sempre é a mesma. Entenda as camadas (unitárias, integração e sistema). E veja com seu chefe qual a expectativa ao criar esses testes.

Outra coisa importante, lembre-se você tem que testar de forma inteligente, e não fazer "testes de melancia onde por fora é verde, mas por dentro vermelho". Em outras palavras tá passando no teste mas o que ele se propõe a testar é divergente do teste em si.

E por último muito cuidado com quantos % vai se comprometer a testar, o [Kent Beck](https://pt.wikipedia.org/wiki/Kent_Beck) o criador do TDD fala o seguinte: "I get paid for code that works, not for tests". Então teste o que realmente é importante e sensível, lembre-se que você é pago para resolver problemas hehe. E antes que se assuste, todas essas siglas relacionadas a teste são metodologias, não existe melhor e nem pior, existe a que te atende!

Espero tê-lo ajudado e abraço!

**Comment by fernandomoraes** [comment]
@edmolima 
Só correção sobre DDD, pois não é "metodologia" de testes.

**Comment by ghost** [comment]
Valeu @fernandomoraes acabou escapando hehe, já corrigi lá mas valeu por ter avisado =)

Só pra documentar o DDD é Desenvolvimento Guiado por Design ou Dominio! E realmente não é uma metodologia para teste!

**Comment by AbraaoAlves** [comment]
É ótimo ver que pessoas como @edmolima cintando o que considero como "A FONTE".

_explicando:_
Toda pergunta quer uma resposta, mas onde busca-la nesse mar de conteúdo é o verdadeira desafio. Muitos podem citar cursos online, presencial, posts, wikipedia, .... Mas no final, você verá que as definições que você quer aprender vêm de um grupo relativamente pequeno.

#### _O que é e onde achar essa FONTE?_  
Respondo com uma pergunta: quem foram os responsáveis por começar e/ou popularizar assuntos sobre as boas praticas de desenvolvimento de software? 

Uma pista:

![Manifesto Agil](https://www.agilealliance.org/wp-content/uploads/2016/06/agile-manifesto-snowbird-1024x768.jpg)
> [THE SOURCE](http://www.manifestoagil.com.br/). Siga o link e veja no rodapé a lista dos mestres

Nos livros desses caras é onde você poderá encontrará a fonte do conteúdo dos cursos, posts e muitos outros assuntos. Apesar de serem escritos em linguagens perdidas no tempo como java(#brinks), seu conteúdo e princípios continuam bem atuais.

#### _Ok, mas preciso de uma sugestão de onde começar_

[Refactoring](https://www.google.com.br/search?sa=X&espv=2&biw=1242&bih=602&q=refactoring+improving+the+design+of+existing+code&stick=H4sIAAAAAAAAAONgFuLQz9U3MLEwNVPiArGM8kqyzYu1pLKTrfST8vOz9RNLSzLyi6xA7GKF_LycSgDsQRntNAAAAA&npsic=-1024&ved=0ahUKEwjH7tfsqurRAhUCgpAKHdyZCy0Q-BYIMg) de [Kent Beck](https://pt.wikipedia.org/wiki/Kent_Beck) é uma ótima pedida para seu dilema, apesar de gostar mais dos livros do [TioBob](https://www.google.com.br/search?sourceid=chrome-psyapi2&ion=1&espv=2&ie=UTF-8&q=uncle%20bob&oq=uncle%20bob&aqs=chrome..69i57j0l5.1714j0j7) e do [careca da foto acima](https://www.google.com.br/search?espv=2&q=Martin+Fowler&stick=H4sIAAAAAAAAAONgFuLWz9U3MDSqMspOT1LiAHFMjSrTtPgCUouK8_OCM1NSyxMriwGlsGhwKQAAAA&sa=X&ved=0ahUKEwjSp6v8qOrRAhWCiZAKHWJjBZEQxA0IswEwFg&biw=1242&bih=602)

Se coisa rápida é o que você quer, veja o capitulo 16 do livro [CleanCode](https://www.google.com.br/search?espv=2&biw=1242&bih=602&q=cleancode&oq=cleancode&gs_l=serp.3..0i7i30k1l2j0i10k1j0i7i30k1l7.5123.5123.0.5442.1.1.0.0.0.0.201.201.2-1.1.0....0...1c.1.64.serp..0.1.199.QHss0Rbf9hE). Lá você encontrará um exemplo passo-a-passo de como refatorar uma biblioteca enquanto faz a cobertura de testes das regras importantes para ela. 
[Não custa avisar]:  esse é o penúltimo capitulo de um livro que não é de bolso, os 15 capítulos que o antecedem lhe darão base para você mesmo fazer refactory com segurança.
